### PR TITLE
Curibe/resiliency score

### PR DESCRIPTION
**Pull request by cristhianu**
Adding ResiliencyScore class to be used by Resiliency Score PDF reports.
Also added Report to RenderingType enum under Rendering.cs

### CODE REVIEW

File: src/Diagnostics.ModelsAndUtils/Models/Rendering.cs [hunk 1]
```diff
@@ -303,7 +303,12 @@ public enum RenderingType
         /// <summary>
         /// StepViews component of Network Troubleshooter
         /// </summary>
-        StepViews
+        StepViews,
+
+        /// <summary>
+        /// Data Rendered as AppInsights enablement in detector
```
Comment (ShekharGupta1988) [inline src/Diagnostics.ModelsAndUtils/Models/Rendering.cs:309]: nit: fix the comment

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        /// Array containing a list of all the resources to be included in the report along with their 
+        /// </summary>
+        ResiliencyResource[] GetResiliencyResourceList();
+
+        /// <summary>
+        /// Array containing a list of all the resources to be included in the report along with their 
+        /// </summary>
+        void SetResiliencyResourceList(ResiliencyResource[] value);
+
+        //      public JObject JSONObject();
```
Comment (rekhaswaminathan) [inline src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs:24]: nit: do you want to remove this comment?

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        /// Comments to be included with the solution in case of a failing grade
+        /// </summary>
+        string SolutionComments { get; set; }
+    }
+
+    public enum Weight
+    {
+        NotCounted = 0,
+        GoodToHave = 1,
+        Important = 5,
```
Comment (rekhaswaminathan) [inline src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs:99]: any reason why its 0,1,5,25 ?

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+		/// </summary>
+		public string SolutionComments { get; set; }
+	}
+
+	public static class ResponseResiliencyReportExtension
+	{
+		public static DiagnosticData AddResiliencyReportData(this Response response, ResiliencyReportData resiliencyReportData)
+		{
+			if (response == null || resiliencyReportData == null)
+            {
```
Comment (rekhaswaminathan) [inline src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs:197]: nit: fix indentation here.

**Review by ShekharGupta1988 — APPROVED** [review]

File: src/Diagnostics.ModelsAndUtils/Models/Rendering.cs (hunk 1 above)
Comment (cristhianu) [reply src/Diagnostics.ModelsAndUtils/Models/Rendering.cs:309]: Thank you for catching this. Fixed now.

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs (hunk 2 above)
Comment (cristhianu) [reply src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs:24]: Thank you for this. Fixed it.

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs (hunk 3 above)
Comment (cristhianu) [reply src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs:99]: When calculating the Resiliency score we wanted to give a weight to some features that was significant enough for those features we consider critical, so we decided to go with that scale.

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs (hunk 4 above)
Comment (cristhianu) [reply src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs:197]: Done. Thank you!

**Review by rekhaswaminathan — APPROVED** [review]

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+{
+    public interface IResiliencyReportData
+    {
+        /// <summary>
+        /// Customer's name used for the report's cover. This will normally be either customer's Company or simply Customer's name obtained from the subscription.
+        /// </summary>
+        string CustomerName { get; set; }
+
+        /// <summary>
+        /// Array containing a list of all the resources to be included in the report along with their
```
Comment (hforeste) [inline src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs:15]: nitpik; finish description

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        /// </summary>
+        string CustomerName { get; set; }
+
+        /// <summary>
+        /// Array containing a list of all the resources to be included in the report along with their 
+        /// </summary>
+        ResiliencyResource[] GetResiliencyResourceList();
+
+        /// <summary>
+        /// Array containing a list of all the resources to be included in the report along with their
```
Comment (hforeste) [inline src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs:20]: nitpik; finish description

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        /// <summary>
+        /// Retrieve the list of Resiliency Features
+        /// </summary>
+        ResiliencyFeature[] GetResiliencyFeaturesList();
+
+
+        /// <summary>
+        /// Set the list of Resiliency Features
+        /// </summary>
+        void SetResiliencyFeaturesList(ResiliencyFeature[] value);
```
Comment (hforeste) [inline src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs:54]: You can sum this up like so:

```suggestion
        /// <summary>
        /// Retrieve the list of Resiliency Features
        /// </summary>
        public ResiliencyFeature[] FeaturesList {get; set;}
```

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+        [JsonConverter(typeof(ResiliencyResource))]
+        ResiliencyResource[] resiliencyResourceList;
+
+        /// <summary>
+        /// Creates an instance of ResiliencyReportData
+        /// </summary>
+        public ResiliencyReportData(string customerName, ResiliencyResource[] resiliencyResourceList)
+        {
+            this.CustomerName = customerName;
+            if (resiliencyResourceList == null)
```
Comment (hforeste) [inline src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs:21]: What about an empty list?

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+    {
+        [JsonConverter(typeof(ResiliencyResource))]
+        ResiliencyResource[] resiliencyResourceList;
+
+        /// <summary>
+        /// Creates an instance of ResiliencyReportData
+        /// </summary>
+        public ResiliencyReportData(string customerName, ResiliencyResource[] resiliencyResourceList)
+        {
+            this.CustomerName = customerName;
```
Comment (hforeste) [inline src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs:20]: What if customer name is null or empty?

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+        /// <summary>
+        /// Customer's name used for the report's cover. This will normally be either customer's Company or simply Customer's name obtained from the subscription.
+        /// </summary>
+        public string CustomerName { get; set; }
+
+        /// <summary>
+        /// Array containing a list of all the resources to be included in the report along with their 
+        /// </summary>
+
+
```
Comment (hforeste) [inline src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs:39]: nitpik; whitespace gap & finish description

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+            if (value == null)
+            {
+                throw new ArgumentNullException(nameof(value), $"{nameof(value)} cannot be null");
+            }
+            else
+            {
+                this.resiliencyResourceList = value;
+            }
+
+        }
```
Comment (hforeste) [inline src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs:56]:
```suggestion
        public ResiliencyResource[] ResiliencyResourceList
        {
             get {
                 return this.resiliencyResourceList;
             }
             
             set {
                 if (value == null)
            {
                throw new ArgumentNullException(nameof(value), $"{nameof(value)} cannot be null");
            }
            else
            {
                this.resiliencyResourceList = value;
            }
             }
        }
```

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+        double _overallScore;
+        ResiliencyFeature[] resiliencyFeaturesList;
+        IDictionary<string, Weight> _featuresDictionary;
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        public ResiliencyResource(string name, IDictionary<string, Weight> featuresDictionary)
+        {
+            this.Name = name;
```
Comment (hforeste) [inline src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs:75]: What if name is null or empty?

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+
+        public ResiliencyFeature[] GetResiliencyFeaturesList()
+        {
+            return this.resiliencyFeaturesList;
+        }
+
+        public void SetResiliencyFeaturesList(ResiliencyFeature[] value)
+        {
+            this.resiliencyFeaturesList = value;
+        }
```
Comment (hforeste) [inline src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs:135]:
```suggestion
        public ResiliencyFeature[] ResiliencyFeaturesList {get; set;}
```

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+
+        /// <summary>
+        /// Call CalculatedOverallScore to recalculate the OverallScore to make sure it calculates with latest changes.
+        /// </summary>
+        public double OverallScore { get => CalculateOverallScore(); }
```
Comment (hforeste) [inline src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs:140]:
```suggestion
        public double OverallScore {
        get {
        _overallScore = 0;
            double _overallScoreSum = 0;
            double _featureWeightsSum = 0;
            foreach (ResiliencyFeature rf in resiliencyFeaturesList)
            {
                Weight _weight;
                Grade _grade;
                bool weightParsingResult = Enum.TryParse<Weight>(rf.FeatureWeight.ToString(), out _weight);
                bool gradeParsingResult = Enum.TryParse<Grade>(rf.ImplementationGrade.ToString(), out _grade);
                if (weightParsingResult && gradeParsingResult)
                {
                    _overallScoreSum += (int)rf.FeatureWeight * (int)rf.ImplementationGrade;
                    _featureWeightsSum += (int)rf.FeatureWeight;
                }
            }
            _overallScore = Math.Round((_overallScoreSum * 100) / (_featureWeightsSum * 2), 1);
            return Math.Round(_overallScore, 2);
        }
        }
```

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+    /// ResiliencyResources supported can be Web App or ASE
+    /// </summary>
+    public class ResiliencyResource : IResiliencyResource
+    {
+        double _overallScore;
+        ResiliencyFeature[] resiliencyFeaturesList;
+        IDictionary<string, Weight> _featuresDictionary;
+
+        /// <summary>
+        /// Constructor
```
Comment (hforeste) [inline src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs:71]: Add something more descriptive

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+
+        /// <summary>
+        /// Comments to be included with the solution in case of a failing grade
+        /// </summary>
+        public string SolutionComments { get; set; }
+    }
+
+    public static class ResponseResiliencyReportExtension
+    {
+        public static DiagnosticData AddResiliencyReportData(this Response response, ResiliencyReportData resiliencyReportData)
```
Comment (hforeste) [inline src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs:198]: Add summary comment above

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs (hunk 5 above)
Comment (cristhianu) [reply src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs:15]: Done. Thank you for catching it.

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs (hunk 6 above)
Comment (cristhianu) [reply src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs:20]: Done. Thank you for catching it, too.

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs (hunk 7 above)
Comment (cristhianu) [reply src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/IResiliencyReportData.cs:54]: That's what I did initially but when I do I get the following:
CA1819: Properties should not return arrays.
So I decided to fix it by using this (in the implementation not in the interface):
https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1819?view=vs-2019#change-the-property-to-a-method

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs (hunk 9 above)
Comment (cristhianu) [reply src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs:20]: Thank you! Changes it to:
            if (string.IsNullOrEmpty(customerName)) 
            {
                throw new ArgumentNullException(nameof(customerName), $"{nameof(customerName)} cannot be null or empty");
            }
            else
            {
                this.CustomerName = customerName;
            }

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs (hunk 8 above)
Comment (cristhianu) [reply src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs:21]: Thanks. Changed it to this:

            if (resiliencyResourceList == null ||  resiliencyResourceList.Length == 0)

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs (hunk 10 above)
Comment (cristhianu) [reply src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs:46]: Done. Thank you

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs (hunk 12 above)
Comment (cristhianu) [reply src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs:82]: Fixed. Thank you

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs (hunk 11 above)
Comment (cristhianu) [reply src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs]: Same case as https://github.com/Azure/Azure-AppServices-Diagnostics/pull/733#discussion_r709591682

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs (hunk 10 above)
Comment (cristhianu) [reply src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs:46]: Fixed. Thank you.

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs (hunk 13 above)
Comment (cristhianu) [reply src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs]: Same case as https://github.com/Azure/Azure-AppServices-Diagnostics/pull/733#discussion_r709526691

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs (hunk 15 above)
Comment (cristhianu) [reply src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs]: Done. Thank you for catching this.

File: src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs (hunk 14 above)
Comment (cristhianu) [reply src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs:175]: Thank you! Great suggestion!

Comment (cristhianu) [reply src/Diagnostics.ModelsAndUtils/Models/ResponseExtensions/ResiliencyReportData.cs:175]: Done.

**Pull request merged**